### PR TITLE
Fixbug return error in handlePackageHeader for the first time

### PR DESCRIPTION
**Pull request by Haiss2**

### CODE REVIEW

File: pkg/multicast/multicast.go [hunk 1]
```diff
@@ -312,21 +324,25 @@ func (c *Client) closeConnection(addr string) error {
...
-func (c *Client) restartConnection(ctx context.Context, addr string) error {
-	conn := c.getConnection(addr)
-	err := conn.Close()
-	if err != nil {
-		c.log.Errorw("failed to close connection", "err", err)
-		return err
-	}
-
-	err = c.buildInstrumentsMapping()
+func (c *Client) restartConnections(ctx context.Context) error {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go:316]: re-initialize `resetConnCh` channel

File: pkg/multicast/multicast.go [hunk 2]
```diff
@@ -391,16 +407,13 @@ func (c *Client) handlePackageHeader(r io.Reader) error {
 	if !ok {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Don't need this check, only need to check if `seq` was set and it's correct.

File: pkg/multicast/multicast.go [hunk 3]
```diff
@@ -517,15 +531,14 @@ func (c *Client) ListenToEventsForAddress(ctx context.Context, addr string) erro
 				err := c.Handle(m, bufferData)
 				if errors.Is(err, ErrConnectionReset) || errors.Is(err, ErrLostPackage) {
 					l.Infow("connection reset or lost package err, restarting connection...", "error", err)
-					err := c.restartConnection(ctx, addr)
-					if err != nil {
-						l.Errorw("fail to restart connection", "error", err)
-					}
-					return
+					c.restartConnCh <- struct{}{}
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: close this channel and then exit this go-routine.

File: pkg/multicast/multicast.go [hunk 4]
```diff
@@ -517,15 +531,14 @@ func (c *Client) ListenToEventsForAddress(ctx context.Context, addr string) erro
...
 				} else if err != nil {
 					l.Errorw("fail to handle UDP package", "error", err)
 				}
+			case <-closeCh:
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Should check for `restartConnCh` was closed and then exit the go-routine.

File: pkg/multicast/multicast.go [hunk 5]
```diff
@@ -536,6 +549,7 @@ func (c *Client) ListenToEventsForAddress(ctx context.Context, addr string) erro
 		if err != nil {
 			if isNetConnClosedErr(err) {
 				l.Infow("connection closed", "error", err)
+				closeCh <- struct{}{}
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: close the data channel and exit, don't need closeCh.

File: pkg/multicast/multicast.go [hunk 6]
```diff
@@ -517,15 +531,14 @@ func (c *Client) ListenToEventsForAddress(ctx context.Context, addr string) erro
 				err := c.Handle(m, bufferData)
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go:516]: Check for `dataCh` to be closed and exit.

File: pkg/multicast/multicast.go [hunk 7]
```diff
@@ -312,21 +324,25 @@ func (c *Client) closeConnection(addr string) error {
...
 		c.log.Errorw("failed to build instruments mapping", "err", err)
 		return err
 	}
 
-	go c.ListenToEventsForAddress(ctx, addr)
+	c.mu.Lock()
+	c.channelIDSeqNum = make(map[uint16]uint32)
+	c.mu.Unlock()
+
+	for addr, conn := range c.addrConnMap {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Also need to delete the connection from map after close it.

File: pkg/multicast/multicast.go (hunk 4 above)
Comment (hiepnv90) [reply pkg/multicast/multicast.go]: Assign `restartConnCh` to a local variable for checking.

File: pkg/multicast/multicast.go [hunk 8]
```diff
@@ -298,6 +300,16 @@ func (c *Client) Start(ctx context.Context) error {
 		return err
 	}
 
+	go func() {
```
Comment (hiepnv90) [inline pkg/multicast/multicast.go]: Don't need a go-routine.

File: pkg/multicast/multicast.go (hunk 1 above)
Comment (Haiss2) [reply pkg/multicast/multicast.go:316]: > 

We don't use this channel anymore.

**Review by hiepnv90 — APPROVED** [review]
LGTM! :+1:

**Pull request merged**